sql-plugin/src/main/scala/com/nvidia/spark/rapids/spill/SpillFramework.scala

```diff
@@ -0,0 +1,1606 @@
+                iter.foreach { bb =>
+                  try {
+                    while (bb.hasRemaining) {
+                      outputChannel.write(bb)
+                    }
+                  } finally {
+                    RapidsStorageUtils.dispose(bb)
+                  }
+                }
+              }
+            }
+            disk = Some(diskHandleBuilder.build)
+            disk
+          }
+        } else {
+          None
+        }
+      }
+      releaseHostResource()
+      sizeInBytes
+    }
+  }
+
+  private def releaseHostResource(): Unit = {
+    SpillFramework.removeFromHostStore(this)
+    synchronized {
+      host.foreach(_.close())
+      host = None
+    }
+  }
+  
+  private def getHostBuffer: Option[HostMemoryBuffer] = synchronized {
+    host.foreach(_.incRefCount())
+    host
+  }
+
+  override def close(): Unit = {
+    releaseHostResource()
+    synchronized {
+      disk.foreach(_.close())
+      disk = None
+    }
+  }
+
+  private[spill] def materializeToDeviceMemoryBuffer(dmb: DeviceMemoryBuffer): Unit = {
+    var hostBuffer: HostMemoryBuffer = null
+    var diskHandle: DiskHandle = null
+    synchronized {
+      if (host.isDefined) {
+        hostBuffer = host.get
+        hostBuffer.incRefCount()
+      } else if (disk.isDefined) {
+        diskHandle = disk.get
+      } else {
+        throw new IllegalStateException(
+          "attempting to materialize a closed handle")
+      }
+    }
+    if (hostBuffer != null) {
+      GpuTaskMetrics.get.readSpillFromHostTime {
+        withResource(hostBuffer) { _ =>
+          dmb.copyFromHostBuffer(
+            /*dstOffset*/ 0,
+            /*src*/ hostBuffer,
+            /*srcOffset*/ 0,
+            /*length*/ hostBuffer.getLength)
+        }
+      }
+    } else {
+      // cannot find a full host buffer, get chunked api
+      // from disk
+      diskHandle.materializeToDeviceMemoryBuffer(dmb)
+    }
+  }
+
+  private[spill] def setHost(singleShotBuffer: HostMemoryBuffer): Unit = synchronized {
+    host = Some(singleShotBuffer)
+  }
+
+  private[spill] def setDisk(handle: DiskHandle): Unit = synchronized {
+    disk = Some(handle)
+  }
+}
+
+object SpillableDeviceBufferHandle {
+  def apply(dmb: DeviceMemoryBuffer): SpillableDeviceBufferHandle = {
+    val handle = new SpillableDeviceBufferHandle(dmb.getLength, dev = Some(dmb))
+    SpillFramework.stores.deviceStore.track(handle)
+    handle
+  }
+}
+
+class SpillableDeviceBufferHandle private (
+    override val sizeInBytes: Long,
+    private[spill] override var dev: Option[DeviceMemoryBuffer],
+    private[spill] var host: Option[SpillableHostBufferHandle] = None)
+    extends DeviceSpillableHandle[DeviceMemoryBuffer] {
+
+  private[spill] override def spillable: Boolean = synchronized {
+    if (super.spillable) {
+      dev.getOrElse {
+        throw new IllegalStateException(
+          s"$this is spillable but it doesn't have a dev buffer!")
+      }.getRefCount == 1
```

Comment:
This logic is basically identical between device and host. Should we consider making some shared templated implementation to dedup this?